include/kodlab_mjbots_sdk/pid.h

```diff
@@ -0,0 +1,192 @@
+/**
+ * @file pid.h
+ * @author Chandravaran Kunjeti (kunjeti@seas.upenn.edu)
+ * @brief Class implementation of PID
+ * @date 9/30/2022
+ *
+ * @copyright 2022 The Trustees of the University of Pennsylvania. All rights reserved.
+ *
+ */
+
+#pragma once
+
+#ifndef _PID_
+#define _PID_
+
+#include "Eigen/Geometry"
+
+namespace kodlab
+{
+
+  /**
+   * @brief Pid object
+   */
+  class Pid
+  {
+
+  public:
+    /**
+     * @brief Construct a Pid object
+     * @param p_gain Proportional Gain
+     * @param d_gain Derivative Gain
+     * @param i_gain Integral Gain
+     * @param time_step Time difference between 2 updates
+     */
+    Pid(double p_gain, double d_gain, double i_gain, double time_step, double deadband, double saturation)
+        : kp_(0), kd_(0), ki_(0), time_step_(0.01), deadband_(0), saturation_(0)
+    {
+
+      kp_ = p_gain;
+      kd_ = d_gain;
+      ki_ = i_gain;
+      time_step_ = time_step;
+      deadband_ = deadband;
+      saturation_ = saturation;
+    }
+
+    /**
+     * @brief Function to set the PID gains
+     * @param p_gain Proportional Gain
+     * @param d_gain Derivative Gain
+     * @param i_gain Integral Gain
+     */
+    void SetGains(double p_gain, double d_gain, double i_gain)
+    {
+
+      kp_ = p_gain;
+      kd_ = d_gain;
+      ki_ = i_gain;
+    }
+
+    /**
+     * @brief Function to set the deadband
+     * @param deadband Deadband that needs to be set
+     */
+    void SetDeadband(double deadband)
+    {
+
+      deadband_ = deadband;
+    }
+
+    /**
+     * @brief Function to set the Saturation
+     * @param saturation Saturation
+     */
+    void SetSaturation(double saturation)
+    {
+
+      saturation_ = saturation;
+    }
+
+    /**
+     * @brief Function to update the PID output.
+     * @param goal Final goal state for the controller.
+     * @param current Current state of the controller.
+     * @return Returns the Pid output
+     */
+    double Update(double goal, double current)
+    {
+
+      // Calculate the errors
+      // Proportional error
+      error_ = goal - current;
+
+      // Integral error
+      i_error_ += error_ * time_step_;
+
+      // Derivative error
+      derivative_ = (error_ - d_error_) / time_step_;
+
+      // Control output
+      output_ = kp_ * error_ + kd_ * derivative_ + ki_ * i_error_;
+
+      // Saturation checking
+      output_ = (output_ > saturation_) ? saturation_ : output_;
+
+      // Storing the past error
+      d_error_ = derivative_;
```

Comment:
Why are all of these computed and saved as class members?  I feel like more information is being retained here than necessary.

Reply:
I was saving this in class variables, so I don't have to reinitialise the variable when I am running the program. It may be minimal but recreating a variable takes up some compute time.

Reply:
I think you are misunderstanding how the compiled code would execute here...

As written, every single computed value is saved as a data member.  As such, each computation step needs to be executed, and the result saved to memory (including values that will not be needed again, like `derivative_` and `output_`). This memory is allotted when the object is initialized and maintained while the object is kept in scope.

Alternately, you could write this function using temporary variables for any data that does not need to be kept.  For example, 
```cpp
float derivative = (error_ - d_error_) / time_step_;
```
When this code is compiled, the compiler will recognize that `derivative` is a temporary variable and (in most cases) optimize it out.  The resulting code will be more efficient at runtime. Although we are only talking about copying a couple of scalar values here, the principle is still important to understand.

Reply:
Okay I understand now.